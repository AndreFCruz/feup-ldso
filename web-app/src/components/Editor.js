import React, {Component} from 'react';
import { Button } from 'reactstrap';
import PropTypes from 'prop-types';

class Editor extends Component {

    constructor(props) {
        super(props);

        this.state = {
            error: null,
            isLoaded: false,
            project: []
        };

        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    componentDidMount() {
        const route = process.env.REACT_APP_ENDPOINT + 'faculties/en/' + this.props.match.params.faculty.toLowerCase() + '/social-projects?filter[where][id]=' + this.props.match.params.project;
        fetch(route)
            .then(res => res.json())
            .then(
                (result) => {
                    this.setState({
                        isLoaded: true,
                        project: result,
                    });
                },
                (error) => {
                    this.setState({
                        isLoaded: true,
                        error
                    });
                }
            );
    }

    handleInputChange(event) {
        this.setState({
            [event.target.name]: event.target.value
        });
    }
    //o url com filtro fica .../social-projects?filter[where][id]=<id_here>

    handleSubmit(event){
        event.preventDefault();
        const route = process.env.REACT_APP_ENDPOINT + 'faculties/en/' + this.props.match.params.faculty.toLowerCase() + '/social-projects?where[id]=' + this.props.match.params.project;
        fetch(route, {
            method: 'PATCH',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(this.state),
        })
            .then(response => response.json());
    }

    render() {
        const { project } = this.state;

        return (
            <div>
                {
                    project.map(proj => (
                        <form key={proj.id}>
                            <label>
          Project Title
                                <br/>
                                <input
                                    name="title"
                                    placeholder={proj.title}
                                    value={this.state.title}
                                    onChange={this.handleInputChange}/>
                            </label>
                            <br/>
                            <label>
          Project Description
                                <br/>
                                <textarea
                                    name="content"
                                    placeholder={proj.content}
                                    value={this.state.content}
                                    onChange={this.handleInputChange}/>
                            </label>
                            <br/>
                            <label>
                                <Button color="secondary" onClick={this.handleSubmit}>Edit</Button>
                            </label>
                        </form>
                    ))}
            </div>
        );}
}

Editor.propTypes = {
    match: PropTypes.object.isRequired
};

export default Editor;
