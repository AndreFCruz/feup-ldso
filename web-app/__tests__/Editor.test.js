import React from 'react';
import renderer from 'react-test-renderer';
import Editor from '../src/components/Editor';
import 'isomorphic-fetch'; //components use fetch() on mount
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';

import Enzyme from 'enzyme';
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

var mockAxios = new MockAdapter(axios);

var props = {
    match: {
        params: {
            faculty: 'faup',
            project: '1'
        }
    }
};

describe('Editor Test Suite', () => {
    it('Edit Snapshot', () => {

        mockAxios.onPatch('http://localhost:3000/faculties/en/faup/social-projects?where[id]=').reply(200, {
            id: 1,
            title: 'Project title',
            content: 'project content',
        });

        const mockupCall = jest.fn();
        const component = renderer.create(
            <Editor
                {...props}
                onChildSetRefresh={mockupCall}
            />
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Submit handler called', () => {
        const mockupCall = jest.fn();
        const wrapper = shallow(<Editor
            {...props}
            onChildSetRefresh={mockupCall}
        />);

        wrapper.setState({
            error: null,
            isLoaded: true,
            project: [{id: 1, title: 'projectToEdit', value: 'valueToEdit'}],
            title: 'TestTitle',
            content: 'This is a mockup test content'
        });

        expect(mockupCall.mock.calls.length).toEqual(0);
        wrapper.find('#submitBtn').simulate('click', {
            preventDefault: () => { }
        });

        expect(mockupCall.mock.calls.length).toEqual(1);
    });

    //TODO: a test to check if project is actually edited
});
