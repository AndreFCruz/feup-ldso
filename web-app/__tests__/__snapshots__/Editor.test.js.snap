// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Test Suite Edit Snapshot 1`] = `<div />`;
